be/src/storage/binlog_builder.h

```diff
@@ -0,0 +1,124 @@
+// Copyright 2021-present StarRocks, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include "storage/binlog_file_writer.h"
+
+namespace starrocks {
+
+// Parameters to create a binlog builder
+struct BinlogBuilderParams {
+    // storage directory for binlog file
+    std::string binlog_storage_path;
+    // maximum size of binlog file
+    int64_t max_file_size;
+    // maximum size of a page in binlog file
+    int32_t max_page_size;
+    // compression type for page
+    CompressionTypePB compression_type;
+    // file id for the next new binlog file
+    int64_t start_file_id;
+    // active file writer and it's current meta. The new ingestion can
+    // append binlog to the writer. nullptr if there is no active writer
+    std::shared_ptr<BinlogFileWriter> active_file_writer;
+    std::shared_ptr<BinlogFileMetaPB> active_file_meta;
+};
+
+// The result of builder which will be applied to the meta of BinlogManager. No matter whether
+// the build is successful (BinlogBuilder#commit) or failed （BinlogBuilder#abort), should always
+// update the metas of BinlogManager, such as next_file_id and active_writer used for the next
+// build. If build successfully also need add new binlog file metas to BinlogManager.
+struct BinlogBuildResult {
+    // parameters used to create the builder
+    std::shared_ptr<BinlogBuilderParams> params;
+    // file id for the next new binlog file in the next ingestion
+    int64_t next_file_id;
+    // the writer can be used for the next build. nullptr if there is
+    // no active writer
+    std::shared_ptr<BinlogFileWriter> active_writer;
+    // New binlog file metas if build successfully
+    std::vector<std::shared_ptr<BinlogFileMetaPB>> metas;
+};
+
+using BinlogBuilderParamsPtr = std::shared_ptr<BinlogBuilderParams>;
+using BinlogBuildResultPtr = std::shared_ptr<BinlogBuildResult>;
+
+// Write the binlog of an ingestion into multiple binlog files.
+//
+// How to use
+//  std::shared_ptr<BinlogBuilder> builder;
+//  // add multiple items
+//  file_writer->add_insert_range()
+//  file_writer->add_insert_range()
+//  // commit the binlog data in this version, and ensure they are persisted
+//  file_writer->commit();
+//  // abort if there is any error when add_xxx() or commit()
```

Comment:
What if aborted? Can it be used after aborting?

Reply:
Yes, the file will be reused if truncated successfully when aborting